### PR TITLE
background-mask-update-speed branch

### PR DESCRIPTION
**Pull request by emasab**
- background mask updated with exponential average to avoid flickering
- used sigmoid for a more clear division between the background mask and the frame

### CODE REVIEW

**Comment by fangfufu** [comment]
The code used to have blurring between the person's face and the background. Mediapipe does the blurring automatically. I have to do more investigation before deciding on whether or not to merge this.

**Comment by emasab** [comment]
Hi nice to meet you. As blurring you're referring to slowing down the high frequency updates of the Mediapipe detected background that cause the flickering, that the exponenial average is trying to address?
Or are you referring to the sigmoid applied to the result from Mediapipe? That it's doing the opposite, making the division sharper

**Comment by fangfufu** [comment]
Have a look at old version of the code:
https://github.com/fangfufu/Linux-Fake-Background-Webcam/blob/bodypix/fakecam/fake.py#L194

Have a look at this blog post
https://elder.dev/posts/open-source-virtual-background/

```python
def post_process_mask(mask):
    mask = cv2.dilate(mask, np.ones((10,10), np.uint8) , iterations=1)
    mask = cv2.blur(mask.astype(float), (30,30))
    return mask
```

These are the blurring operation I was referring to.

File: fake.py [hunk 1]
```diff
@@ -427,6 +438,15 @@ def getNextOddNumber(number):
         return number + 1
     return number
 
+def getPercentage(number):
+    return min(max(int(number), 0), 100)
```
Comment (fangfufu) [inline fake.py:442]: Surely we can just use ``round`` here?

**Review by fangfufu — CHANGES_REQUESTED** [review]
I would like to have a flag to completely turn this off.

**Comment by emasab** [comment]
Thanks for the clarification, I've tried it but I still find the result with the sigmoid less "ghosty" than with the dilate and blur or than without  additional processing. Also, I'd like if the noise reduction with the exponential average wouldn't come at the cost of a slower mask update but it seems that the output of mediapipe is a bit noisy

File: fake.py (hunk 1 above)
Comment (emasab) [reply fake.py:442]: Here I'm letting the user input a number between 0 and 100 and limiting it to 0 or 100 if it's lower or higher.

Comment (fangfufu) [reply fake.py:442]: I suppose it is better to sanitise user input like what you did here.

File: fake.py [hunk 2]
```diff
@@ -445,7 +465,8 @@ def main():
         foreground_mask_image=findFile(args.foreground_mask_image, args.image_folder),
         webcam_path=args.webcam_path,
         v4l2loopback_path=args.v4l2loopback_path,
-        ondemand=not args.no_ondemand)
+        ondemand=not args.no_ondemand,
+        background_mask_update_speed=getPercentage(args.background_mask_update_speed))
```
Comment (fangfufu) [inline fake.py:469]: I would like a flag to completely turn this off.

**Review by fangfufu — CHANGES_REQUESTED** [review]

File: fake.py (hunk 2 above)
Comment (emasab) [reply fake.py:469]: just set it to 100

Comment (fangfufu) [reply fake.py:469]: Line 276 still runs if you set it to 100. This extra processing can drop FPS on low end machines.

Comment (emasab) [reply fake.py:469]: done

File: fake.py [hunk 3]
```diff
@@ -413,6 +422,8 @@ def parse_args():
                         help="Add a hologram effect")
     parser.add_argument("--no-ondemand", action="store_true",
                         help="Continue processing when no consumers are present")
+    parser.add_argument("--background-mask-update-speed", default="10", type=int,
```
Comment (fangfufu) [inline fake.py:425]: I don't like the lag with the default update speed. Should we make it go faster by default?

Comment (emasab) [reply fake.py:425]: what's the default you are thinking to?

Comment (fangfufu) [reply fake.py:425]: 30 to 50 maybe?

**Comment by fangfufu** [comment]
By the way, have you tried thresholding? That should help with flickering too.

File: fake.py (hunk 3 above)
Comment (emasab) [reply fake.py]: ok

**Comment by fangfufu** [comment]
Thanks, merging it in! :)

**Comment by emasab** [comment]
tried now with 
`mask = self.classifier.process(frame).segmentation_mask > 0.5`
but it doesn't help much with flickering

**Comment by fangfufu** [comment]
I will have a look at thresholding. The old bodypix version used thresholding to create a more stable mask. I just thought with the new neural network, I don't have to do anything.

**Pull request merged**